### PR TITLE
[FLINK-35461][config] Improve Runtime Configuration for Flink 2.0

### PR DESCRIPTION
**Pull request by Sxnan**
## What is the purpose of the change

As Flink moves toward 2.0, we have revisited all runtime configurations and identified several improvements to enhance user-friendliness and maintainability. We want to refine the runtime configuration.

## Brief change log

- Deprecate options related to hash-based blocking shuffle
- Deprecate options related to legacy hybrid shuffle
- Deprecate or update options of floating/exclusive buffer
- Deprecate or update options of shuffle compression
- Deprecate netty options that are not recommended to be used
- Deprecate taskmanager.network.max-num-tcp-connections
- Deprecate the unnecessary option fine-grained.shuffle-mode.all-blocking
- Deprecate StreamPipelineOptions that only contains deprecated option

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by Sxnan** [comment]
@flinkbot run azure

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java [hunk 1]
```diff
@@ -428,6 +455,21 @@ public enum CompressionCodec {
                                     + "tasks have occupied all the buffers and the downstream tasks are waiting for the exclusive buffers. The timeout breaks"
                                     + "the tie by failing the request of exclusive buffers and ask users to increase the number of total buffers.");
 
+    /** The timeout for requesting buffers for each channel. */
+    @Documentation.ExcludeFromDocumentation(
+            "This option is purely implementation related, and may be removed as the implementation changes.")
+    public static final ConfigOption<Duration> NETWORK_BUFFERS_REQUEST_TIMEOUT =
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:473]: The old `ConfigOption` can be removed as it is already covered by `withDeprecatedKeys` of the new option.

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java [hunk 2]
```diff
@@ -316,7 +316,11 @@ public enum CompressionCodec {
     /**
      * Parallelism threshold to switch between sort-based blocking shuffle and hash-based blocking
      * shuffle.
+     *
+     * @deprecated The hash-based blocking shuffle is deprecated in 1.20 and will be totally removed
+     *     in 2.0.
      */
+    @Deprecated
     @Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:344]: Same for other deprecated options with documentation related annotations.

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java (hunk 1 above)
Comment (Sxnan) [reply flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:462]: Removing the old `ConfigOption` right away will cause the code that directly uses the `ConfigOption` to fail. It is better to mark it deprecated in 1.20 and remove it in 2.0.

**Review by TanYuxin-tyx — COMMENTED** [review]
Thanks @Sxnan for the contribution. I left some minor comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NettyShuffleEnvironmentConfiguration.java [hunk 3]
```diff
@@ -373,11 +373,18 @@ public static NettyShuffleEnvironmentConfiguration fromConfiguration(
         BoundedBlockingSubpartitionType blockingSubpartitionType =
                 getBlockingSubpartitionType(configuration);
 
-        boolean batchShuffleCompressionEnabled =
-                configuration.get(NettyShuffleEnvironmentOptions.BATCH_SHUFFLE_COMPRESSION_ENABLED);
         CompressionCodec compressionCodec =
                 configuration.get(NettyShuffleEnvironmentOptions.SHUFFLE_COMPRESSION_CODEC);
 
+        boolean batchShuffleCompressionEnabled;
+        if (compressionCodec == CompressionCodec.NONE) {
```
[the remark this answers is not in the document]
Comment (Sxnan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NettyShuffleEnvironmentConfiguration.java:380]: The `taskmanager.network.batch-shuffle.compression.enabled` is `true` by default, the new user that only  `taskmanager.network.batch-shuffle.compression.enabled` to `true` will always see the warning and have to set explicitly set the `taskmanager.network.batch-shuffle.compression.enabled` to `false`, if we print a warning in this case.

However, I think it is reasonable to print a warning if the user set `taskmanager.network.batch-shuffle.compression.enabled` to `false` to disable the compression

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java [hunk 4]
```diff
@@ -425,6 +446,21 @@ public enum CompressionCodec {
                                     + "tasks have occupied all the buffers and the downstream tasks are waiting for the exclusive buffers. The timeout breaks"
                                     + "the tie by failing the request of exclusive buffers and ask users to increase the number of total buffers.");
 
+    /** The timeout for requesting buffers for each channel. */
+    @Documentation.ExcludeFromDocumentation(
+            "This option is purely implementation related, and may be removed as the implementation changes.")
+    public static final ConfigOption<Duration> NETWORK_BUFFERS_REQUEST_TIMEOUT =
+            key("taskmanager.network.memory.exclusive-buffers-request-timeout")
```
[the remark this answers is not in the document]
Comment (Sxnan) [reply flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java]: Good catch!

**Comment by Sxnan** [comment]
@xintongsong @TanYuxin-tyx Thanks for the review! I updated the PR accordingly. Please take another look.

**Review by TanYuxin-tyx — APPROVED** [review]
@Sxnan Thanks for the update. I have no more comments on the change.

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java (hunk 1 above)
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:462]: Ok, make sense.

**Review by xintongsong — APPROVED** [review]
LGTM. Merging.

**Pull request closed without merging**